### ISSUE TITLE
Questions about reproducing results on COCO

### ISSUE THREAD

**Issue by GhostWnd** [Alibaba-MIIL/ASL#36]
Hello, I tried to reproduce the result on COCO, I implemented my own framework and most of my files are the same as you, I only write my new train.py.
As is introduced in your paper, I have implemented EMA with 0.999, 1cycle policy with max learing rate 2e-4, Adam optimizer with weigth_decay 1e-4, img_size = 448*448 and batchsize = 16.

But when I train my model, the loss decreases from 120 to around 90 and then it just doesn't decrease and the performance on validation date is very bad, whose mAP is around 10, at first I guess it's because I didn't spend much time training, I only trained it for an hour, but when I try to train it longer, the loss still doesn't decrease, could you please tell me what I have done wrong?

My code is avaliable at https://github.com/GhostWnd/reproducingASL, thank you for your help.

**Comment by mrT23** [comment]
three observations to start with:

for testing convergence, use smaller resolution (224) and larger batch size (128)

i don't see any augmentations in your training code.
use RandAugment or AutoAugment at least + cutout

also, something weird with your scheduler:
`    scheduler = lr_scheduler.OneCycleLR(optimizer, max_lr = 0.0002, total_steps = total_step, epochs = 25)
`
its hardcoded to 25 epochs, yet you loop over only 5 epochs
add these hyperparameters to arg list, and use them everywhere instead of hard-coded numbers:

**Comment by mrT23** [comment]
also for testing and prototyping, use tresnet_m

**Comment by GhostWnd** [comment]
Thanks for your comment, I will try and report later.

**Comment by mrT23** [comment]
i will take a look at the code and try to run It when i have the time.

good work so far, i think with joint forces we are on our way to finally have a modern multi-label code for the community to use, that vast majority of repo that exist are way outdated.

several more corrections and suggestions and corrections:

1) args.do_bottleneck_head = False (not True)

2)
one more correction is that you are using the 2017 split. while this is not a "mistake" (and your results will be a little higher), in articles people use the 2014 split.

3) what about mixed precision ? with modern pytorch it is a few line of code  ("with autocast():"...)

4) to improve speed, you don't have to update EMA every iteration. you cant update it every ~5 iterations with slightly higher decaey rate, and still ger similar results.

5) load a pretrained  model, run only inference, and make sure you reproduce the article results (after switching to 2014 split)

6) make sure, especially in validation, that you are not building enormous vectors along the training that clog the RAM memory. sometimes its better to pre-allocate memory if you need to store large vectors

**Comment by GhostWnd** [comment]
Thank you for your comment, I will correct it.
And if it's possible, could you please release the loss change when you run my code?
Thank you very much.

**Comment by mrT23** [comment]
Hi GhostWnd

i took a deeper look at the code. there are several major problems there. 
make sure you understand whats the problem in each and every one of them, and apply corrections. don't skip a single one. 
most of these problems can be a "deal breaker". and its also a good journey to understand what it takes to create a top repo.


- currently not using randaugment (commented)

- not initilizaing model (from relevant imagenet model https://github.com/Alibaba-MIIL/TResNet/blob/master/MODEL_ZOO.md)

- doing "target = target.max(dim=1)[0]". remove this line

- using 2017 coco split (instead 2014 coco split )

-  Cutout(n_holes = 1, length = 16)  ->  Cutout(n_holes = 1, args.image_size/2)

- validation should be done once an epoch, no more and no less

- 
        preds.append(output.cpu())
        targets.append(target.cpu())
		
		->
	preds.append(output.cpu().detach())
        targets.append(target.cpu().detach())
		
		
-
mAP_score = validate_multi(val_loader,  model, args, ema)
->
model.eval()
mAP_score = validate_multi(val_loader, model, args, ema)
model.train()
				
- calculate only mAP metrics. remove other metrics, they are only confusing during training

**Comment by mrT23** [comment]
just to give you motivation, i got a good score last night when running a corrected code...

**Comment by GhostWnd** [comment]
Thank you for your comment and effort, I will try to correct the code and run it.
Thank you very much.

**Comment by GhostWnd** [comment]
Sorry to bother again, I know due to commercial issues you can't release your training code, but could you release the code you correct based on my train.py? 

If that is not possible, could you please release the loss record of your corrected code based on my training code, so that I can compare the result myself?

Thank you very much.

**Comment by mrT23** [comment]
Hi  GhostWnd 

there were other problems in the code.
The two major ones:
- sigmoid was done twice (!) - once in the direct prediction, second in the loss.
- EMA was not performed correctly (its a separate model with separate validation)

anyway, this code fully reproduces  the article results:
[train_asl_reproduce.zip](https://github.com/Alibaba-MIIL/ASL/files/5917657/train_asl_reproduce.zip)

i will attach logs for 224 and 448 trainings later

you are welcome to test it yourself and give me feedback.

thanks for the collaboration, together we will release the first publicly available modern multi-label code
:-)

**Comment by GhostWnd** [comment]
Thank you so much!  :-)

**Comment by mrT23** [comment]
this is my log file:
[mtresnet_224.txt](https://github.com/Alibaba-MIIL/ASL/files/5918781/mtresnet_224.txt)

**Comment by mrT23** [comment]
> Thank you so much! :-)
> I will upload the train file to make it publicly available as well as test it by myself and give feedback to you as soon as possible.

do you have an objection that i will add the code also to
https://github.com/Alibaba-MIIL/ASL ?
i think it will help it gain more traction. there are very few (zero) modern multi-label code-bases like this, with top results.

i will of course share credit with you, i had made a lot of changes and enhancements to the code, but you provided the base implementation.

**Comment by GhostWnd** [comment]
No objection, it's my pleasure, thank you very much.

**Comment by GhostWnd** [comment]
And I wonder whether you can put the mode based on tresnet_m and input size 224 into you pretrained model?

I would like to apply it to other dataset as pretrained model.
Thank you very much.

**Comment by mrT23** [comment]
i am not sure i fully understand your question.

models in
https://github.com/Alibaba-MIIL/ASL/blob/main/MODEL_ZOO.md
are standard imagenet models for downstream tasks. these are the models you should use to initialize training on COCO.

**Comment by GhostWnd** [comment]
Well, I think you make a little mistake, 
models in https://github.com/Alibaba-MIIL/ASL/blob/**main**/MODEL_ZOO.md are models trained on MSCOCO
while models in https://github.com/Alibaba-MIIL/ASL/blob/**master**/MODEL_ZOO.md are standard imagenet models, right?

I just wonder whether you could upload the model you trained with tresnet_m with input size 224 into https://github.com/Alibaba-MIIL/ASL/blob/**main**/MODEL_ZOO.md

**Comment by GhostWnd** [comment]
Or could you please share the model you trained on tresnet_m and input size 224 with me?

I would like to adjust some hyper parameters to test the influence of those hyperparameters.
And apply it to other dataset as pretrained model.
Thank you very much.

**Comment by mrT23** [comment]
just to be clear:
tresnet_m 224 model trained on MS-COCO ?

**Comment by GhostWnd** [comment]
yes, the one that produces the log file mtresnet_224.txt

**Comment by mrT23** [comment]
added to
https://github.com/Alibaba-MIIL/ASL/blob/main/MODEL_ZOO.md

**Comment by GhostWnd** [comment]
Thank you very much.

**Comment by LOOKCC** [comment]
> this is an example log file (notice - resolution 224, mtresnet)
> [mtresnet_224.txt](https://github.com/Alibaba-MIIL/ASL/files/5918781/mtresnet_224.txt)

Can you attach logs for 448 resolution with tresnet_l using this training code? I found it's hard for me to reprodect the 86.8mAP resault in paper.

**Comment by mrT23** [comment]
@LOOKCC 
run 
https://github.com/Alibaba-MIIL/ASL/blob/main/train.py
